### PR TITLE
ENHANCEMENT: ONLIGHT CHANGES

### PR DESCRIPTION
**Pull request by dweeezil**
We've been using ldap2dns for a few years and encountered some problems building and using your latest version.  This patch should fix all the issues and will also add Debian build support.
- Add a "-i userid" option
  - Fix 32 vs. 64 bit operations and some typos.
  - Added STRUCTURAL to dnszone, dnsrrset and dnsloccodes.
  - Remove the "SUP dnszone" from the "dnsrrset" objectclass.
  - Added the dnslocation attribute to the dnsrrset object class.
  - Make sure the update interval is set in the case of invokation as "ldap2dnsd"
    when using foreground daemon mode.
  - Add return-value checking for the nice() and the system() calls.  This should
    eliminate all compilation warnings.
  - Add Debian packaging.

### CODE REVIEW

**Comment by bklang** [comment]
Awesome, thanks!

File: Makefile [hunk 1]
```diff
@@ -1,5 +1,6 @@
-# $Id$ 
+# $Id: Makefile 418 2008-01-21 20:39:55Z bklang $ 
 VERSION=0.5.0
+ONLVERSION=$Revision: 14040 $
```
Comment (bklang) [inline Makefile]: What is the purpose of this variable? I don't see it referenced in the Makefile.

Incidentally, it looks like the $Id$ SVN tags were expanded at some point in your checkout. Since we're not longer using SVN I think it makes sense to remove them entirely, as they will never change once committed.

Comment (dweeezil) [inline Makefile]: On 08/28/2013 11:30 AM, Ben Klang wrote:

> In Makefile:
> 
> >  VERSION=0.5.0
> > +ONLVERSION=$Revision: 14040 $
> 
> What is the purpose of this variable? I don't see it referenced in the 
> Makefile.
> 
> Incidentally, it looks like the $Id$ SVN tags were expanded at some 
> point in your checkout. Since we're not longer using SVN I think it 
> makes sense to remove them entirely, as they will never change once 
> committed.
> 
> Grrumble, I was just actually about to send you a note saying that 
> you'll likely want to blow away the SVN tags.

The ONLVERSION was a booboo on my part.  I forgot to remove it.  In our 
internal version, we alter the printed-out version number to have an 
Onlight (our company) suffix so that when we install it on a customer's 
system, we can tell that it's using our version.

If you'd like, I can re-roll the commit but I think you can probably 
just as easily clean these things out.

```
 - Tim
```

Comment (bklang) [inline Makefile]: Actually, you don't even need to re-roll the commit, just make one more commit to your branch pulling out these tags.  It'll automatically update the pull request.

Comment (dweeezil) [inline Makefile]: Yes, of course.  I've gotten so used to some of the other projects where the maintainers want all their pull requests to appear as a single commit in order to minimize the history.  I'll prepare a cleaned up commit shortly.

File: debian/ldap2dns-doc.debhelper.log [hunk 2]
```diff
@@ -0,0 +1,44 @@
+dh_installman
```
Comment (bklang) [inline debian/ldap2dns-doc.debhelper.log]: Should this log file really be checked in?

File: debian/ldap2dns.debhelper.log [hunk 3]
```diff
@@ -0,0 +1,44 @@
+dh_installman
```
Comment (bklang) [inline debian/ldap2dns.debhelper.log]: Ditto: should this log file really be checked in?

File: debian/postinst [hunk 4]
```diff
@@ -0,0 +1,43 @@
+#! /bin/sh
+# postinst script for ldap2dns
+
+set -e
+
+case "$1" in
+    configure)
+	cd /usr/bin
```
Comment (bklang) [inline debian/postinst]: This might seem pedantic, but would you mind fixing the whitespace to be at least consistent internally? For shell scripts I don't have a strong preference of tabs vs. spaces, but it should be consistent throughout for readability.

File: debian/postrm [hunk 5]
```diff
@@ -0,0 +1,23 @@
+#! /bin/sh
+# postrm script for ldap2dns
+
+set -e
+
+case "$1" in
+       purge|upgrade|failed-upgrade|abort-upgrade)
+       ;;
+	   remove|abort-install|disappear)
```
Comment (bklang) [inline debian/postrm]: Ditto: whitespace

File: debian/prerm [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+#        * <conflictor's-prerm> `remove' `in-favour' <package> <new-version>
+#        * <deconfigured's-prerm> `deconfigure' `in-favour'
+#          <package-being-installed> <version> `removing'
+#          <conflicting-package> <version>
+# for details, see /usr/share/doc/packaging-manual/
+
+case "$1" in
+    remove|upgrade|deconfigure)
+	
+	for i in `ls -d /var/lib/svscan/ldap2tinydns* 2>/dev/null`; do
```
Comment (bklang) [inline debian/prerm]: Ditto: whitespace

File: debian/ldap2dns-doc.debhelper.log (hunk 2 above)
Comment (dweeezil) [inline debian/ldap2dns-doc.debhelper.log]: Of course not!  My co-worker prepared our internal changes (in our svn repo) and clearly left these in accidentally.  I'll get rid of them in the commit I'm about to push.

File: ldap2dns.schema [hunk 7]
```diff
@@ -140,6 +140,7 @@ attributetype ( 1.3.6.1.4.1.7222.1.4.27
 
 objectclass ( 1.3.6.1.4.1.7222.1.4.19
 	NAME 'dnszone'
+	STRUCTURAL
```
Comment (bklang) [inline ldap2dns.schema:142]: Why did you change this to be STRUCTURAL? Same with 'dnsrrset' below

File: ldap2dns.xml [hunk 8]
```diff
@@ -1,24 +1,11 @@
...
-
-This program is distributed in the hope that it will be useful,
-but WITHOUT ANY WARRANTY; without even the implied warranty of
-MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-GNU General Public License for more details.
-
-You should have received a copy of the GNU General Public License
-along with this program; if not, write to the Free Software
-Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+	Copyright 2004 Sun Microsystems, Inc.  All rights reserved.
```
Comment (bklang) [inline ldap2dns.xml:4]: Why did the copyright change here?

File: scripts/data2ldif.pl [hunk 9]
```diff
@@ -354,7 +354,7 @@
             print $outfh "cn: $id\n";
             print $outfh "dnstype: txt\n";
             print $outfh "dnsdomainname: $fqdn.\n";
-            print $outfh "dnstxt: $s\n";
+            print $outfh "dnscname: $s.\n";
```
Comment (bklang) [inline scripts/data2ldif.pl]: This reverts a change I made. It appears you may be working from an old checkout.

Since this worries me a bit (not sure I caught all the changes in ldap2dns.c), perhaps you can rebase your changes off the latest from Git?

File: ldap2dns.schema (hunk 7 above)
Comment (dweeezil) [inline ldap2dns.schema:142]: That's something my co-worker will have to answer as he knows a whole lot more about LDAP than I do.  I'll ask him to answer this when he gets back into the office.

**Comment by dweeezil** [comment]
@bklang With respect to the reversion that occurred in the Perl script, I'd like to explain the heritage of our code and the rationale for the way we prepared this commit.

Our original intent was to start out with your latest git master and retrofit our fixes.  Unfortunately, between the last release (0.4.2?) and your current git master, a number of bugs slipped into the program so we decided instead to take our current internally-modified code and retrofit your 0.4.2-to-0.5.0 changes into it.  Apparently we missed the change to the Perl script.

Once we had the code tested and deployed in a few places, we decided it would be a good idea to push these changes to you (finally!).  That required converting our internal code base as-is to a patch on top of your git master and that's what I originally submitted.

**Comment by bklang** [comment]
Ok, that rationale makes sense.  Have all the bugs you identified in the 0.4.2 -> 0.5.0 upgrade been addressed by this commit? If not, would you open issues so they can at least be tracked?

Once this is merged, do you plan to continue running from Git (or at least a branch from this Git master)?

**Comment by dweeezil** [comment]
Yes, they have all been addressed.

Once merged, we'd like to stop using our internal svn version because your 0.5.0 (on github) would be considered the canonical version.  At the very least, we'd start using an internal git repository (although our main corporate repository is svn, we do have a number of git repos, too).

Also, on an unrelated note, I'll deal with unifying the whitespace.  My co-worker deals with the Debian-ization and I suspect he has various bits and pieces of scripts that are pieced together as appropriate.  Also, he uses an editor (geany) that, I think, tries some high-order magic when dealing with the whole space-versus-tab issue.

File: ldap2dns.schema (hunk 7 above)
Comment (postilion) [inline ldap2dns.schema:142]: According to RFC2252, section 4.4 (https://www.ietf.org/rfc/rfc2252.txt):

```
4.4. Object Classes

   The format for representation of object classes is defined in X.501
   [3]. In general every entry will contain an abstract class ("top" or
   "alias"), at least one structural object class, and zero or more
   auxiliary object classes.
```

Since the dnsZone, dnsLocation and dnsRRset objectClasses are designed to permit the creation of directory entries, they must either be STRUCTURAL or else be AUXILIARY to another STRUCTURAL objectClass.  The latter doesn't make practical sense, so the former must be so.

This is fairly well described in "LDAP for Rocket Scientists" (http://www.zytrax.com/books/ldap/ch3/#overview):
- objectClasses group sets of attributes:
  -    objectclasses are defined inside schemas.
  -    objectclasses may be organised in a hierarchy in which case they inherit all the properties of their parents or SUPerior.
  -  objectclasses may be STRUCTURAL, in which case they are used to create entries (data objects), AUXILIARY in which case they may be added into any convenient entry, or ABSTRACT - a non-existent 'thingie'. The most common ABSTRACT objectclass is top, which forms the highest level of every objectclass hierarchy, and terminates any hierarchy.

Also, in your original schema, you had declared the dnsRRset objectClass inferior to dnsZone, like so:

```
objectclass ( 1.3.6.1.4.1.7222.1.4.20
  NAME 'dnsrrset'
  SUP dnszone
  MUST ( objectclass $ cn )
  MAY ( dnsdomainname $ dnsrr $ dnsclass $ dnstype $ dnsipaddr
    $ dnscipaddr $ dnscname $ dnspreference $ dnsttl $ dnstimestamp
    $ owner $ dnssrvpriority $ dnssrvweight $ dnssrvport ) )
```

The "SUP dnszone" declaration means that this objectClass would inherit all attributes of the SUPerior objectClass, including NIChandle and the like.  This cannot be what was intended.
